bundles/org.openhab.binding.androidtv/src/main/java/org/openhab/binding/androidtv/internal/protocol/googletv/GoogleTVConnectionManager.java

```diff
@@ -0,0 +1,1383 @@
+                senderThread.setDaemon(true);
+                senderThread.start();
+                this.senderThread = senderThread;
+
+                logger.trace("{} - Checking for PIN MODE for {}:{} {}", handler.getThingID(), config.ipAddress,
+                        config.port, config.mode);
+
+                if (config.mode.equals(PIN_MODE)) {
+                    logger.trace("{} - Sending PIN Login to {}:{}", handler.getThingID(), config.ipAddress,
+                            config.port);
+                    // Send app name and device name
+                    sendCommand(new GoogleTVCommand(GoogleTVRequest.encodeMessage(GoogleTVRequest.loginRequest(1))));
+                    // Unknown but required
+                    sendCommand(new GoogleTVCommand(GoogleTVRequest.encodeMessage(GoogleTVRequest.loginRequest(2))));
+                    // Don't send pin request yet, let user send REQUEST via PINCODE channel
+                } else {
+                    logger.trace("{} - Not PIN Mode {}:{} {}", handler.getThingID(), config.ipAddress, config.port,
+                            config.mode);
+                }
+            } else {
+                scheduleConnectRetry(config.reconnect); // Possibly a temporary problem. Try again later.
+            }
+        }
+    }
+
+    public void shimInitialize() {
+        synchronized (connectionLock) {
+            AndroidTVPKI shimPKI = new AndroidTVPKI();
+            byte[] shimEncryptionKey = shimPKI.generateEncryptionKey();
+            SSLContext sslContext;
+
+            try {
+                shimPKI.generateNewKeyPair(shimEncryptionKey);
+                // Move this to PKI. Shim requires a trusted cert chain in the keystore.
+                KeyStore keystore = KeyStore.getInstance("JKS");
+                FileInputStream keystoreInputStream = new FileInputStream(config.keystoreFileName);
+                keystore.load(keystoreInputStream, config.keystorePassword.toCharArray());
+
+                KeyManagerFactory kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+                kmf.init(keystore, config.keystorePassword.toCharArray());
+                TrustManager[] trustManagers = defineNoOpTrustManager();
+
+                sslContext = SSLContext.getInstance("TLS");
+                sslContext.init(kmf.getKeyManagers(), trustManagers, null);
+                this.sslServerSocketFactory = sslContext.getServerSocketFactory();
+
+                logger.trace("Opening GoogleTV shim on port {}", config.port);
+                SSLServerSocket sslServerSocket = (SSLServerSocket) this.sslServerSocketFactory
+                        .createServerSocket(config.port);
+                if (this.config.mode.equals(DEFAULT_MODE)) {
+                    sslServerSocket.setNeedClientAuth(true);
+                } else {
+                    sslServerSocket.setWantClientAuth(true);
+                }
+
+                while (true) {
```

Comment:
I think it is reusing a common openHAB thread pool? If so, it would be better to reschedule a new job when exception occur, so the thread is returned to the pool and can be used by other openHAB add-ons.

Reply:
This is looping not due to an exception but to wait for the next connection.  In theory when we hit line 708 cleanly (meaning the connection was good and nothing failed, we go back to line 644 to wait for the next connection.  Also, shim is a developer item for me to do break and inspect on the app, it is NOT enabled by the end users so this should never impact actual user operations.